diff --git a/dwl.c b/dwl.c
index fbb5e81..f873792 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1319,7 +1319,7 @@ void
 mapnotify(struct wl_listener *listener, void *data)
 {
 	/* Called when the surface is mapped, or ready to display on-screen. */
-	Client *c = wl_container_of(listener, c, map);
+	Client *c = wl_container_of(listener, c, map), *oldfocus = selclient();
 
 	if (client_is_unmanaged(c)) {
 		/* Insert this independent into independents lists. */
@@ -1328,9 +1328,15 @@ mapnotify(struct wl_listener *listener, void *data)
 	}
 
 	/* Insert this client into client lists. */
-	wl_list_insert(&clients, &c->link);
-	wl_list_insert(&fstack, &c->flink);
-	wl_list_insert(&stack, &c->slink);
+	if (oldfocus) {
+		wl_list_insert(oldfocus->link.prev, &c->link);
+		wl_list_insert(oldfocus->flink.prev, &c->flink);
+		wl_list_insert(oldfocus->slink.prev, &c->slink);
+	} else {
+		wl_list_insert(&clients, &c->link);
+		wl_list_insert(&fstack, &c->flink);
+		wl_list_insert(&stack, &c->slink);
+	}
 
 	client_get_geometry(c, &c->geom);
 	c->geom.width += 2 * c->bw;
