diff --git a/config.mk b/config.mk
index cd4e821..9a886cc 100644
--- a/config.mk
+++ b/config.mk
@@ -5,4 +5,4 @@ PREFIX = /usr/local
 CFLAGS ?= -g -Wall -Wextra -Werror -Wno-unused-parameter -Wno-sign-compare -Wno-unused-function -Wno-unused-variable -Wdeclaration-after-statement
 
 # Uncomment to build XWayland support
-#CFLAGS += -DXWAYLAND
+CFLAGS += -DXWAYLAND
diff --git a/scripts/guix-install.sh b/scripts/guix-install.sh
index b6d17b1..039548c 100755
--- a/scripts/guix-install.sh
+++ b/scripts/guix-install.sh
@@ -1,2 +1,2 @@
 #!/bin/sh
-guix environment dwl --ad-hoc pkg-config wlroots guile -- make clean dwl CC=gcc
+guix environment dwl --ad-hoc pkg-config wlroots guile xorg-server-xwayland -- make clean dwl CC=gcc
