diff --git a/dwl.c b/dwl.c
index 2e56ed2..32e3ce9 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1559,8 +1559,13 @@ mapnotify(struct wl_listener *listener, void *data)
         Client *c = wl_container_of(listener, c, map), *sel = selclient();
 
         /* Insert this client into client lists. */
-        wl_list_insert(&clients, &c->link);
-        wl_list_insert(&fstack, &c->flink);
+        if (sel) {
+                wl_list_insert(sel->link.prev, &c->link);
+                wl_list_insert(sel->flink.prev, &c->flink);
+        } else {
+                wl_list_insert(&clients, &c->link);
+                wl_list_insert(&fstack, &c->flink);
+        }
 
         client_get_geometry(c, &c->geom);
         c->geom.width += 2 * c->bw;
