diff --git a/dwl.c b/dwl.c
index 2b75de9..c1b163b 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1663,7 +1663,7 @@ void
 mapnotify(struct wl_listener *listener, void *data)
 {
 	/* Called when the surface is mapped, or ready to display on-screen. */
-	Client *p, *w, *c = wl_container_of(listener, c, map);
+	Client *p, *w, *c = wl_container_of(listener, c, map), *sel = focustop(selmon);
 	Monitor *m;
 	int i;
 
@@ -1709,8 +1709,13 @@ mapnotify(struct wl_listener *listener, void *data)
 	c->geom.height += 2 * c->bw;
 
 	/* Insert this client into client lists. */
-	wl_list_insert(&clients, &c->link);
-	wl_list_insert(&fstack, &c->flink);
+	if (sel) {
+		wl_list_insert(sel->link.prev, &c->link);
+		wl_list_insert(sel->flink.prev, &c->flink);
+	} else {
+		wl_list_insert(&clients, &c->link);
+		wl_list_insert(&fstack, &c->flink);
+	}
 
 	/* Set initial monitor, tags, floating status, and focus:
 	 * we always consider floating, clients that have parent and thus
