diff --git a/dwl.c b/dwl.c
index 0ef7a35..47a96a0 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1432,11 +1432,16 @@ void
 mapnotify(struct wl_listener *listener, void *data)
 {
 	/* Called when the surface is mapped, or ready to display on-screen. */
-	Client *c = wl_container_of(listener, c, map);
+	Client *c = wl_container_of(listener, c, map), *oldfocus = selclient();
 
 	/* Insert this client into client lists. */
-	wl_list_insert(&clients, &c->link);
-	wl_list_insert(&fstack, &c->flink);
+	if (oldfocus) {
+		wl_list_insert(oldfocus->link.prev, &c->link);
+		wl_list_insert(oldfocus->flink.prev, &c->flink);
+	} else {
+		wl_list_insert(&clients, &c->link);
+		wl_list_insert(&fstack, &c->flink);
+	}
 
 	client_get_geometry(c, &c->geom);
 	c->geom.width += 2 * c->bw;
