diff --git a/config.mk b/config.mk
index 7669208..f57df46 100644
--- a/config.mk
+++ b/config.mk
@@ -8,8 +8,5 @@ MANDIR = $(PREFIX)/share/man
 CFLAGS ?= -g -Wall -Wextra -Werror -Wno-unused-parameter -Wno-sign-compare -Wno-unused-function -Wno-unused-variable -Wno-unused-result -Wdeclaration-after-statement
 
 DEBUGFLAGS =
-XWAYLAND =
-XLIBS =
-# Uncomment to build XWayland support
-#XWAYLAND = -DXWAYLAND
-#XLIBS = xcb xcb-icccm
+XWAYLAND = -DXWAYLAND
+XLIBS = xcb xcb-icccm
diff --git a/guix.scm b/guix.scm
index aea3569..daf449f 100644
--- a/guix.scm
+++ b/guix.scm
@@ -131,7 +131,9 @@
   (inputs
    (modify-inputs (package-inputs dwl)
                   (prepend guile-3.0
-                           wayland-1.21.0)
+                           wayland-1.21.0
+                           xorg-server-xwayland-22.1.5
+                           xcb-util-wm)
                   (replace "wlroots" wlroots-0.16.0)))
   (arguments
     (substitute-keyword-arguments
