diff --git a/config.mk b/config.mk
index cee96fa..c860870 100644
--- a/config.mk
+++ b/config.mk
@@ -7,8 +7,8 @@ MANDIR = $(PREFIX)/share/man
 # Default compile flags (overridable by environment)
 CFLAGS ?= -g -Wall -Wextra -Werror -Wno-unused-parameter -Wno-sign-compare -Wno-unused-function -Wno-unused-variable -Wno-unused-result -Wdeclaration-after-statement
 
-XWAYLAND =
-XLIBS =
+# XWAYLAND =
+# XLIBS =
 # Uncomment to build XWayland support
-#XWAYLAND = -DXWAYLAND
-#XLIBS = xcb xcb-icccm
+XWAYLAND = -DXWAYLAND
+XLIBS = xcb xcb-icccm
diff --git a/guix.scm b/guix.scm
index ad493f2..49cdd55 100644
--- a/guix.scm
+++ b/guix.scm
@@ -132,7 +132,9 @@
   (inputs
    (modify-inputs (package-inputs dwl)
                   (prepend guile-3.0
-                           wayland-1.21.0)
+                           wayland-1.21.0
+                           xorg-server-xwayland-22.1.5
+                           xcb-util-wm)
                   (replace "wlroots" wlroots-0.16.0)))
   (arguments
     (substitute-keyword-arguments
